Add paintCurrentVideoFrameInContext method to AudioVideoRenderer

jyavenard opened the pull request:
#### 5f3fff28b7dcce320e18c4d05ab8295d31a041d6
<pre>
Add paintCurrentVideoFrameInContext method to AudioVideoRenderer
<a href="https://bugs.webkit.org/show_bug.cgi?id=299419">https://bugs.webkit.org/show_bug.cgi?id=299419</a>
<a href="https://rdar.apple.com/161227994">rdar://161227994</a>

Reviewed by Eric Carlson.

The MediaPlayerPrivateWebM and MediaPlayerPrivateMediaSourceAVFObjC only knew
how to paint in a GraphicsContext a native image.

When running in the web content process, this isn&apos;t a functional codepath.

We need to add support for dealing with RemoteVideoFrameProxy.

Added test to exercise the new codepath.

Test: media/video-webm-canvas-drawing.html
* LayoutTests/media/content/counting.webm: Added.
* LayoutTests/media/video-webm-canvas-drawing-expected.txt: Added.
* LayoutTests/media/video-webm-canvas-drawing.html: Added.
* Source/WebCore/platform/graphics/AudioVideoRenderer.h:
(WebCore::VideoInterface::paintCurrentVideoFrameInContext):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::isCrossOrigin const):
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
(WebCore::AudioVideoRendererAVFObjC::paintCurrentVideoFrameInContext):
(WebCore::AudioVideoRendererAVFObjC::setVideoFullscreenLayer):
(WebCore::AudioVideoRendererAVFObjC::currentNativeImage):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::updateLastVideoFrame):
(WebCore::MediaPlayerPrivateWebM::paintCurrentFrameInContext):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::paintCurrentVideoFrameInContext):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:

Canonical link: <a href="https://commits.webkit.org/300819@main">https://commits.webkit.org/300819@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f60c193230b2f1400e032a32df14252a2d50842

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123956 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43671 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34368 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130747 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44395 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52265 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94282 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9bbf15dc-bd3d-4f77-8750-79120886c238) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35373 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/110885 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74883 "Found 2 new API test failures: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlot, TestWTF:WTF_Condition.OneProducerOneConsumerOneSlotTimeout (failure)") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8ad076f9-5996-4c05-80ee-52de7b7dc4b7) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34321 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29047 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74219 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105104 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29269 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133441 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50909 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38776 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102752 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51283 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107105 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102576 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26092 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47923 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26176 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47762 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50762 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56526 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50236 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53582 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51910 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/MediaStrategy.cpp:
```diff
@@ -83,4 +87,30 @@ void MediaStrategy::addMockMediaSourceEngine()
+void MediaStrategy::enableRemoteRenderer(MediaPlayerMediaEngineIdentifier type, bool enabled)
+{
+    if (enabled)
+    m_remoteRenderersEnabled |= 1 << static_cast<uint32_t>(type);
```
Nit: should be indented

eric-carlson commented on Source/WebCore/platform/MediaStrategy.cpp:
```diff
@@ -83,4 +87,30 @@ void MediaStrategy::addMockMediaSourceEngine()
+
+bool MediaStrategy::hasRemoteRendererFor(MediaPlayerMediaEngineIdentifier type) const
+{
+    return m_remoteRenderersEnabled & 1 << static_cast<uint32_t>(type);
```
Can `m_remoteRenderersEnabled` be an `OptionSet< MediaPlayerMediaEngineIdentifier >` instead of a `uint32_t`  so you don't need to `static_cast`?

jyavenard replied:
MediaPlayerMediaEngineIdentifier is currently an enum, if I was to change it for one usable in an OptionSet it would no longer work when sending over IPC. As it must be a plain enum or an OptionSet; it can't be both with the current serialiser.

jyavenard committed:
Add paintCurrentVideoFrameInContext method to AudioVideoRenderer
https://bugs.webkit.org/show_bug.cgi?id=299419
rdar://161227994

Reviewed by Eric Carlson.

The MediaPlayerPrivateWebM and MediaPlayerPrivateMediaSourceAVFObjC only knew
how to paint in a GraphicsContext a native image.

When running in the web content process, this isn't a functional codepath.

We need to add support for dealing with RemoteVideoFrameProxy.

Added test to exercise the new codepath.

Test: media/video-webm-canvas-drawing.html
* LayoutTests/media/content/counting.webm: Added.
* LayoutTests/media/video-webm-canvas-drawing-expected.txt: Added.
* LayoutTests/media/video-webm-canvas-drawing.html: Added.
* Source/WebCore/platform/graphics/AudioVideoRenderer.h:
(WebCore::VideoInterface::paintCurrentVideoFrameInContext):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::isCrossOrigin const):
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
(WebCore::AudioVideoRendererAVFObjC::paintCurrentVideoFrameInContext):
(WebCore::AudioVideoRendererAVFObjC::setVideoFullscreenLayer):
(WebCore::AudioVideoRendererAVFObjC::currentNativeImage):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::updateLastVideoFrame):
(WebCore::MediaPlayerPrivateWebM::paintCurrentFrameInContext):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::paintCurrentVideoFrameInContext):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:

Canonical link: https://commits.webkit.org/300819@main
